toolchain/base/canonical_value_store.h

```diff
@@ -70,37 +72,43 @@ class CanonicalValueStore {
   Set<IdT, /*SmallSize=*/0, KeyContext> set_;
 };
 
-template <typename IdT, typename KeyT, typename ValueT>
-class CanonicalValueStore<IdT, KeyT, ValueT>::KeyContext
+template <typename IdT, typename KeyT, typename ValueT, auto ValueToKeyFn>
+class CanonicalValueStore<IdT, KeyT, ValueT, ValueToKeyFn>::KeyContext
     : public TranslatingKeyContext<KeyContext> {
  public:
   explicit KeyContext(const ValueStore<IdT, ValueType>* values)
       : values_(values) {}
 
   // Note that it is safe to return a `const` reference here as the underlying
-  // object's lifetime is provided by the `ValueStore`.
-  auto TranslateKey(IdT id) const -> ConstRefType { return values_->Get(id); }
+  // object's lifetime is provided by the `ValueStore`. Returns `auto` to pick
+  // up the return type of `ValueToKeyFn`.
+  auto TranslateKey(IdT id) const -> auto {
```

Comment:
Consider returning `decltype(auto)` instead so that a reference return type is properly forwarded.

Reply:
Are you okay with `llvm::function_traits<decltype(ValueToKeyFn)>::result_t` here? Mainly avoids the need for comments.